{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "689f55f6-d0af-4576-9922-efed2c964e5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load library\n",
    "library(tximport)\n",
    "library(tidyverse)\n",
    "library(biomaRt)\n",
    "library(DESeq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13b0ea8f-dec8-43a3-bffe-8f0e45509791",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m9\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m3\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (3): sample, sra_accession, group\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### Step1_Tximport RHAM1 ####\n",
    "### Tximport ###\n",
    "# load study design file to load Kallisto\n",
    "targets <- read_tsv(\"Sequencing_data/kallisto/studydesign.txt\")\n",
    "path <- file.path(\"Sequencing_data/kallisto\", targets$sample, \"abundance.tsv\") \n",
    "all(file.exists(path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "493f247c-0bca-434c-baf3-246aaf10b515",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Ensembl will soon enforce the use of https.\n",
      "Ensure the 'host' argument includes \"https://\"”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>biomart</th><th scope=col>version</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>plants_mart      </td><td>Ensembl Plants Genes 60     </td></tr>\n",
       "\t<tr><td>plants_variations</td><td>Ensembl Plants Variations 60</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " biomart & version\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t plants\\_mart       & Ensembl Plants Genes 60     \\\\\n",
       "\t plants\\_variations & Ensembl Plants Variations 60\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| biomart &lt;chr&gt; | version &lt;chr&gt; |\n",
       "|---|---|\n",
       "| plants_mart       | Ensembl Plants Genes 60      |\n",
       "| plants_variations | Ensembl Plants Variations 60 |\n",
       "\n"
      ],
      "text/plain": [
       "  biomart           version                     \n",
       "1 plants_mart       Ensembl Plants Genes 60     \n",
       "2 plants_variations Ensembl Plants Variations 60"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Ensembl will soon enforce the use of https.\n",
      "Ensure the 'host' argument includes \"https://\"”\n",
      "Warning message:\n",
      "“Ensembl will soon enforce the use of https.\n",
      "Ensure the 'host' argument includes \"https://\"”\n",
      "Note: importing `abundance.h5` is typically faster than `abundance.tsv`\n",
      "\n",
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "\n",
      "\n",
      "summarizing abundance\n",
      "\n",
      "summarizing counts\n",
      "\n",
      "summarizing length\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# load reference from ensembl\n",
    "listMarts(host=\"plants.ensembl.org\")\n",
    "myMart <- useMart(biomart=\"plants_mart\", host=\"plants.ensembl.org\")\n",
    "available.datasets <- listDatasets(myMart)\n",
    "At.anno <- useMart(biomart=\"plants_mart\", host=\"plants.ensembl.org\",\n",
    "                   dataset = \"athaliana_eg_gene\")\n",
    "At.attributes <- listAttributes(At.anno)\n",
    "Tx.At <- getBM(attributes=c('ensembl_gene_id',\n",
    "                            'external_gene_name'),\n",
    "               mart = At.anno)\n",
    "\n",
    "# make reference tx2gene table\n",
    "Tx.At <- as_tibble(Tx.At)\n",
    "Tx.At <- dplyr::rename(Tx.At, target_id = ensembl_gene_id, \n",
    "                       gene_name = external_gene_name)\n",
    "Tx2Gene.AT <- Tx.At\n",
    "Tx2Gene.AT$gene_name <- Tx.At$target_id \n",
    "\n",
    "# tximport from RG-I kallisto results\n",
    "txi.kallisto.tsv <- tximport(path, \n",
    "                             type = \"kallisto\", \n",
    "                             tx2gene = Tx2Gene.AT, \n",
    "                             txOut = FALSE, #Raw name TRUE: target_id, FALSE: gene_id\n",
    "                             countsFromAbundance = \"no\", \n",
    "                             #if your RNAs are not from 3'end sequencing method, use \"lengthScaledTPM\"\n",
    "                             #countsFromAbundance = \"lengthScaledTPM\",\n",
    "                             ignoreTxVersion = TRUE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "005361e3-af6d-403b-815d-3d17e19a7ae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Setting row names on a tibble is deprecated.”\n",
      "using counts and average transcript lengths from tximport\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### Step2_DESeq RHAM1 ####\n",
    "# Set the experiment design\n",
    "samples.rh <- targets\n",
    "samples.rh$condition <- factor(rep(c(\"A\", \"B\", \"C\"),each=3))\n",
    "rownames(samples.rh) <- samples.rh$sample\n",
    "\n",
    "# DEseq using Tximport with RG-I tsv file\n",
    "ddsrh <- DESeqDataSetFromTximport(txi.kallisto.tsv,\n",
    "                                  colData = samples.rh,\n",
    "                                  design = ~ condition)\n",
    "ddsrh.dds <- DESeq(ddsrh)\n",
    "colnames(ddsrh.dds) <- samples.rh$sample\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9b8370cc-fc89-47f1-a028-7917cf76f8c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"01.DESeq_results/rh_30min_gene.csv\"\n",
      "[1] \"01.DESeq_results/rh_90min_gene.csv\"\n"
     ]
    }
   ],
   "source": [
    "# Get results comparing with RG-I 0min treatment\n",
    "res.30min.rh <- results(ddsrh.dds, contrast=c('condition', 'B', 'A'))\n",
    "res.90min.rh <- results(ddsrh.dds, contrast=c('condition', 'C', 'A'))\n",
    "res.30min.rh.df <- as.data.frame(res.30min.rh)\n",
    "res.90min.rh.df <- as.data.frame(res.90min.rh)\n",
    "res.30min.rh.full <- merge(res.30min.rh.df, as.data.frame(counts(ddsrh.dds, normalized=TRUE)), by=\"row.names\", sort=FALSE)\n",
    "res.90min.rh.full <- merge(res.90min.rh.df, as.data.frame(counts(ddsrh.dds, normalized=TRUE)), by=\"row.names\", sort=FALSE)\n",
    "names(res.30min.rh.full)[1] <- \"Gene\"\n",
    "names(res.90min.rh.full)[1] <- \"Gene\"\n",
    "\n",
    "# Generate RG-I full DESeq results file and export\n",
    "res.30min.rh.full <- res.30min.rh.full %>%\n",
    "  dplyr::arrange(Gene)\n",
    "res.90min.rh.full <- res.90min.rh.full %>%\n",
    "  dplyr::arrange(Gene)\n",
    "\n",
    "write_csv(res.30min.rh.full, \"01.DESeq_results/rh_30min_gene.csv\")\n",
    "print(\"01.DESeq_results/rh_30min_gene.csv\")\n",
    "write_csv(res.90min.rh.full, \"01.DESeq_results/rh_90min_gene.csv\")\n",
    "print(\"01.DESeq_results/rh_90min_gene.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
