{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e089a0e-8881-42cd-a67b-404b41f5a854",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load library\n",
    "library(tximport)\n",
    "library(tidyverse)\n",
    "library(biomaRt)\n",
    "library(DESeq2)\n",
    "library(getDEE2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a610f8b3-48b7-48d9-8a0f-707cffeb8d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>SRR_accession</th><th scope=col>QC_summary</th><th scope=col>SRX_accession</th><th scope=col>SRS_accession</th><th scope=col>SRP_accession</th><th scope=col>Experiment_title</th><th scope=col>GEO_series</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>DRR000618</td><td>FAIL(3,4)  </td><td>DRX000325</td><td>DRS000288</td><td>DRP000209</td><td>Arabidopsis Transcriptome Multiplex-1_No1</td><td></td></tr>\n",
       "\t<tr><th scope=row>2</th><td>DRR000619</td><td>FAIL(3,4,6)</td><td>DRX000326</td><td>DRS000289</td><td>DRP000209</td><td>Arabidopsis Transcriptome Multiplex-1_No3</td><td></td></tr>\n",
       "\t<tr><th scope=row>3</th><td>DRR000620</td><td>FAIL(3,4,6)</td><td>DRX000327</td><td>DRS000290</td><td>DRP000209</td><td>Arabidopsis Transcriptome Multiplex-1_No4</td><td></td></tr>\n",
       "\t<tr><th scope=row>4</th><td>DRR000621</td><td>FAIL(3,4,6)</td><td>DRX000328</td><td>DRS000291</td><td>DRP000209</td><td>Arabidopsis Transcriptome Multiplex-1_No5</td><td></td></tr>\n",
       "\t<tr><th scope=row>5</th><td>DRR008476</td><td>PASS       </td><td>DRX007662</td><td>DRS007600</td><td>DRP001015</td><td>Arabidopsis WT-Col mRNA_seq              </td><td></td></tr>\n",
       "\t<tr><th scope=row>6</th><td>DRR008477</td><td>PASS       </td><td>DRX007663</td><td>DRS007601</td><td>DRP001015</td><td>Arabidopsis ibm1-4 mRNA_seq              </td><td></td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & SRR\\_accession & QC\\_summary & SRX\\_accession & SRS\\_accession & SRP\\_accession & Experiment\\_title & GEO\\_series\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & DRR000618 & FAIL(3,4)   & DRX000325 & DRS000288 & DRP000209 & Arabidopsis Transcriptome Multiplex-1\\_No1 & \\\\\n",
       "\t2 & DRR000619 & FAIL(3,4,6) & DRX000326 & DRS000289 & DRP000209 & Arabidopsis Transcriptome Multiplex-1\\_No3 & \\\\\n",
       "\t3 & DRR000620 & FAIL(3,4,6) & DRX000327 & DRS000290 & DRP000209 & Arabidopsis Transcriptome Multiplex-1\\_No4 & \\\\\n",
       "\t4 & DRR000621 & FAIL(3,4,6) & DRX000328 & DRS000291 & DRP000209 & Arabidopsis Transcriptome Multiplex-1\\_No5 & \\\\\n",
       "\t5 & DRR008476 & PASS        & DRX007662 & DRS007600 & DRP001015 & Arabidopsis WT-Col mRNA\\_seq               & \\\\\n",
       "\t6 & DRR008477 & PASS        & DRX007663 & DRS007601 & DRP001015 & Arabidopsis ibm1-4 mRNA\\_seq               & \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | SRR_accession &lt;chr&gt; | QC_summary &lt;chr&gt; | SRX_accession &lt;chr&gt; | SRS_accession &lt;chr&gt; | SRP_accession &lt;chr&gt; | Experiment_title &lt;chr&gt; | GEO_series &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | DRR000618 | FAIL(3,4)   | DRX000325 | DRS000288 | DRP000209 | Arabidopsis Transcriptome Multiplex-1_No1 | <!----> |\n",
       "| 2 | DRR000619 | FAIL(3,4,6) | DRX000326 | DRS000289 | DRP000209 | Arabidopsis Transcriptome Multiplex-1_No3 | <!----> |\n",
       "| 3 | DRR000620 | FAIL(3,4,6) | DRX000327 | DRS000290 | DRP000209 | Arabidopsis Transcriptome Multiplex-1_No4 | <!----> |\n",
       "| 4 | DRR000621 | FAIL(3,4,6) | DRX000328 | DRS000291 | DRP000209 | Arabidopsis Transcriptome Multiplex-1_No5 | <!----> |\n",
       "| 5 | DRR008476 | PASS        | DRX007662 | DRS007600 | DRP001015 | Arabidopsis WT-Col mRNA_seq               | <!----> |\n",
       "| 6 | DRR008477 | PASS        | DRX007663 | DRS007601 | DRP001015 | Arabidopsis ibm1-4 mRNA_seq               | <!----> |\n",
       "\n"
      ],
      "text/plain": [
       "  SRR_accession QC_summary  SRX_accession SRS_accession SRP_accession\n",
       "1 DRR000618     FAIL(3,4)   DRX000325     DRS000288     DRP000209    \n",
       "2 DRR000619     FAIL(3,4,6) DRX000326     DRS000289     DRP000209    \n",
       "3 DRR000620     FAIL(3,4,6) DRX000327     DRS000290     DRP000209    \n",
       "4 DRR000621     FAIL(3,4,6) DRX000328     DRS000291     DRP000209    \n",
       "5 DRR008476     PASS        DRX007662     DRS007600     DRP001015    \n",
       "6 DRR008477     PASS        DRX007663     DRS007601     DRP001015    \n",
       "  Experiment_title                          GEO_series\n",
       "1 Arabidopsis Transcriptome Multiplex-1_No1           \n",
       "2 Arabidopsis Transcriptome Multiplex-1_No3           \n",
       "3 Arabidopsis Transcriptome Multiplex-1_No4           \n",
       "4 Arabidopsis Transcriptome Multiplex-1_No5           \n",
       "5 Arabidopsis WT-Col mRNA_seq                         \n",
       "6 Arabidopsis ibm1-4 mRNA_seq                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "For more information about DEE2 QC metrics, visit\n",
      "    https://github.com/markziemann/dee2/blob/master/qc/qc_metrics.md\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### Step2_getDEE2 OG ####\n",
    "# Load data from NP\n",
    "samples.og <- read.csv(\"Sequencing_data/NP/NP_ex_design.csv\")\n",
    "samples.og <- samples.og %>%\n",
    "  dplyr::select(Source.Name, Comment.ENA_RUN., Comment.ENA_SAMPLE.,  Characteristics.ecotype., Characteristics.stimulus., Comment.replicate., Factor.Value.time.) %>%\n",
    "  dplyr::filter(Characteristics.stimulus. == \"OGs\") %>%\n",
    "  #dplyr::filter(Comment.replicate. == 1) %>%\n",
    "  dplyr::filter(grepl(pattern = \"000|030|090\",Source.Name)) %>% # to extract 00min 30min 90min dataset\n",
    "  dplyr::distinct() # to get unique data raw\n",
    "# Choose your samples based on GEO or SRA ID\n",
    "mySamples.og <- c(samples.og$Comment.ENA_RUN.)\n",
    "mySamples.og.name <- c(samples.og$Source.Name)\n",
    "mySamples.og.name <- unique(mySamples.og.name)\n",
    "mysample.og.temp <- samples.og \n",
    "colnames(mysample.og.temp)[2] <- \"SRR_accession\" #rename 2nd column name for right_join\n",
    "# Search metadata \n",
    "mdat <- getDEE2Metadata(\"athaliana\")\n",
    "head(mdat)\n",
    "# Since mySamples have SRR_accession info, we need to change them to SRP_accession vector\n",
    "mdat.mySamples.og <- mdat %>% dplyr::filter(SRR_accession %in% mySamples.og) %>% # make a dataframe which have only our samples\n",
    "  right_join(mysample.og.temp, by = \"SRR_accession\") %>% # add information from OG_sample info to arrange\n",
    "  dplyr::filter(!is.na(QC_summary)) %>% # remove NA because these dataset were not include in mdat\n",
    "  dplyr::arrange(Factor.Value.time.) # arrange by treatment time \n",
    "mySamples.og.SRR <- mdat.mySamples.og$SRR_accession # get SRR information\n",
    "SRRvec.og <- as.vector(mySamples.og.SRR) # format change as vector with order\n",
    "\n",
    "#Extract expression data from DEE2\n",
    "expression.og <- getDEE2(\"athaliana\", SRRvec.og, \n",
    "                         metadata = mdat, \n",
    "                         counts = \"Tx2Gene\", \n",
    "                         legacy = TRUE)\n",
    "#species, SRRvec, metadata df, counts method\n",
    "## counts method : \"GeneCounts\" (STAR), \"TxCounts\" (Kallisto), \"Tx2Gene\" (Kallisto count aggregated by sum)\n",
    "## legacy: TRUE -> return list data, default -> return SummarizedExperiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a9603786-b468-455c-82c9-2b9f7eb1a800",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### Step2_DESeq OG ####\n",
    "# Tximport OGs data to DEseq2\n",
    "# Make a matrix of OG treatment TxCounts\n",
    "#input.og <- expression.og$TxCounts\n",
    "#input.og <- round(input.og) #round counts to make integer\n",
    "input.og <- expression.og$GeneCounts\n",
    "\n",
    "# Make sample information\n",
    "sample.og <- mdat.mySamples.og %>%\n",
    "  dplyr::select(\"SRR_accession\", \"Source.Name\", \"Comment.replicate.\", \"Factor.Value.time.\")\n",
    "sample.og$condition <- factor(rep(c(\"A\", \"B\", \"C\"),each=4))\n",
    "sample.og <- sample.og[order(sample.og$SRR_accession),]\n",
    "rownames(sample.og) <- sample.og$SRR_accession #df is ordered by SRR_accession\n",
    "\n",
    "# DEseq using Matrix import with og matrix file: input.og\n",
    "ddsog <- DESeqDataSetFromMatrix(countData=input.og, colData = sample.og, design = ~ condition)\n",
    "ddsog.dds <- DESeq(ddsog)\n",
    "colnames(ddsog.dds) <- sample.og$Source.Name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c570c990-fd01-4fae-912c-42f59625dd7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"01.DESeq_results/og_30min_gene.csv\"\n",
      "[1] \"01.DESeq_results/og_90min_gene.csv\"\n"
     ]
    }
   ],
   "source": [
    "# Get results comparing with og 0min treatment\n",
    "res.30min.og <- results(ddsog.dds, contrast=c('condition', 'B', 'A'))\n",
    "res.90min.og <- results(ddsog.dds, contrast=c('condition', 'C', 'A'))\n",
    "res.30min.og.df <- as.data.frame(res.30min.og)\n",
    "res.90min.og.df <- as.data.frame(res.90min.og)\n",
    "res.30min.og.full <- merge(res.30min.og.df, as.data.frame(counts(ddsog.dds, normalized=TRUE)), by=\"row.names\", sort=FALSE)\n",
    "res.90min.og.full <- merge(res.90min.og.df, as.data.frame(counts(ddsog.dds, normalized=TRUE)), by=\"row.names\", sort=FALSE)\n",
    "names(res.30min.og.full)[1] <- \"Gene\"\n",
    "names(res.90min.og.full)[1] <- \"Gene\"\n",
    "\n",
    "# Rearrange OG file by treatment time\n",
    "res.30min.og.full <- res.30min.og.full %>%\n",
    "  dplyr::select(\"Gene\", \"baseMean\", \"log2FoldChange\", \"lfcSE\", \"stat\", \"pvalue\", \"padj\", \n",
    "                \"R1_Col_OGs_000\", \"R2_Col_OGs_000\", \"R3_Col_OGs_000\", \"R4_Col_OGs_000\",\n",
    "                \"R1_Col_OGs_030\", \"R2_Col_OGs_030\", \"R3_Col_OGs_030\", \"R4_Col_OGs_030\",\n",
    "                \"R1_Col_OGs_090\", \"R2_Col_OGs_090\", \"R3_Col_OGs_090\", \"R4_Col_OGs_090\")\n",
    "res.90min.og.full <- res.90min.og.full %>%\n",
    "  dplyr::select(\"Gene\", \"baseMean\", \"log2FoldChange\", \"lfcSE\", \"stat\", \"pvalue\", \"padj\", \n",
    "                \"R1_Col_OGs_000\", \"R2_Col_OGs_000\", \"R3_Col_OGs_000\", \"R4_Col_OGs_000\",\n",
    "                \"R1_Col_OGs_030\", \"R2_Col_OGs_030\", \"R3_Col_OGs_030\", \"R4_Col_OGs_030\",\n",
    "                \"R1_Col_OGs_090\", \"R2_Col_OGs_090\", \"R3_Col_OGs_090\", \"R4_Col_OGs_090\")\n",
    "\n",
    "# Generate og full DESeq results file and export\n",
    "res.30min.og.full <- res.30min.og.full %>%\n",
    "  dplyr::arrange(Gene)\n",
    "res.90min.og.full <- res.90min.og.full %>%\n",
    "  dplyr::arrange(Gene)\n",
    "#write_csv(res.30min.og.full, \"og_30min.csv\")\n",
    "#write_csv(res.90min.og.full, \"og_90min.csv\")\n",
    "write_csv(res.30min.og.full, \"01.DESeq_results/og_30min_gene.csv\")\n",
    "print(\"01.DESeq_results/og_30min_gene.csv\")\n",
    "write_csv(res.90min.og.full, \"01.DESeq_results/og_90min_gene.csv\")\n",
    "print(\"01.DESeq_results/og_90min_gene.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
